src/lib.rs

```diff
@@ -1,253 +1,253 @@
-        let task_monitor = Arc::clone(&task_monitor);
-        self
-          .thread_pool()
-          .spawn_with_handle(future::lazy(move |context| {
-            trace_scoped!(task.name.clone());
-            let task_id = task_monitor.lock().unwrap().started(task.name.clone());
-            let result = (task.task)(&context);
-            task_monitor.lock().unwrap().finished(task_id);
-            (task.name, result)
-          }))
-          .expect("failed to spawn job")
-      })
-      .collect();
-
-    let handles = futures::executor::block_on(future::join_all(handles));
-    pb.finish();
-
-    let mut successful = Vec::new();
-    let mut failed = Vec::new();
-
-    for (task_name, result) in handles {
-      match result {
-        Ok(result) => {
-          successful.push(ExecutionResult {
-            name: task_name,
-            result,
-          });
-        }
 
-        Err(err) => failed.push(ExecutionResult {
-          name: task_name,
-          result: err,
-        }),
-      }
+    fn thread_pool(&mut self) -> &mut rayon::ThreadPool {
+        if self.thread_pool.is_some() {
+            self.thread_pool.as_mut().unwrap()
+        } else {
+            let builder = rayon::ThreadPoolBuilder::new();
+            let builder = match self.thread_count {
+                Some(count) => builder.num_threads(count),
+                None => builder,
+            };
+
+            self.thread_pool = Some(
+                builder
+                    .build()
+                    .unwrap_or_else(|err| panic!("failed to create job pool: {}", err)),
+            );
+            self.thread_pool.as_mut().unwrap()
+        }
     }
 
-    ExecutionResults { successful, failed }
-  }
+    pub fn execute<T: Send, E: Send>(&mut self, mut job: Job<T, E>) -> ExecutionResults<T, E> {
+        let mut task_monitor = TaskMonitor::new(if self.quiet {
+            indicatif::ProgressBar::hidden()
+        } else {
+            let task_count = job.tasks.len();
+            let pb = indicatif::ProgressBar::new(task_count.try_into().unwrap());
+            pb.set_style(progress_bar_style(task_count));
+            pb.set_prefix(job.name.clone());
+            pb.enable_steady_tick(1000);
+            pb
+        });
+
+        tracing::span!(tracing::Level::INFO, "Pool::execute", name = job.name).in_scope(|| {
+            let mut successful = Vec::new();
+            let mut failed = Vec::new();
+            {
+                let state = Mutex::new((&mut task_monitor, &mut successful, &mut failed));
+                self.thread_pool().scope(|scope| {
+                    for task in job.tasks.drain(..) {
+                        scope.spawn(|_| {
+                            tracing::span!(tracing::Level::INFO, "Task::task", name = task.name)
+                                .in_scope(|| {
+                                    let task_id = {
+                                        let mut guard = state.lock().unwrap();
```

Comment:
For a future PR, we might want to switch over to [parking_log](https://crates.io/crates/parking_lot) crate

Reply:
What for?